docs/0003-how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,136 @@
+indirect dependencies that use the other one, so you need to understand how both
+work. To limit the scope, we ignore other build environments like Eclipse and
+Ant that are not based on the Maven repository system.
+
+Maven and Gradle use different names for the process of selecting dependency
+versions: Maven calls the process *dependency mediation*, while Gradle calls the
+process *version conflict resolution*. In this article, we use the term *version
+resolution*, omitting the word "conflict" so that the process is not confused
+with program-breaking conflicts that we discuss elsewhere. 
+
+Before looking at Maven and Gradle specifically, it's important to understand
+the concepts of version resolution in general. The input of this process is a
+dependency graph which reflects the dependencies as specified by each library in
+the graph, which we call an *unresolved dependency graph*. In such a graph,
+there can be multiple versions of each library. The version resolution process
+walks the unresolved dependency graph and decides which version to use for every
+library encountered. Whether or not the build tool produces a graph as an output
+of this resolution, it is useful to think of this output in terms of a new
+graph. We call the graph representation of this output a *resolved dependency
+graph*. Here is a visual representation of the process:
+
+<img src="assets/images/ddc-resolution-00.png" alt="version resolution process">
+
+For example, we could have an unresolved graph that looks like this:
+
+<img src="assets/images/ddc-resolution-01.png" alt="unresolved graph">
+
+In this graph, there are two versions of A (10.0 and 10.1). The version
+resolution process picks a single version for A, which could be either version
+depending on the algorithm. If A:10.0 is picked, the resolved dependency graph
+looks like the following. In this graph, C's dependency on A:10.1 is overridden.
+
+<img src="assets/images/ddc-resolution-02.png" alt="resolved graph, A:10.0 chosen">
+
+If A:10.1 is picked, the resolved dependency graph looks like the following. In
+this graph, B's dependency on A:10.0 is overridden.
+
+<img src="assets/images/ddc-resolution-03.png" alt="resolved graph, A:10.1 chosen">
+
+So why do we need to pick a single version? The reason is that each class loader
+only loads one version of each fully-qualified class name at runtime. So, if you
+put multiple versions on the classpath, you do not get both versions at
+runtime. Java build systems handle this issue upfront by picking a single
+version for each library before the classpath is even constructed.
+
+Now let's look at the version resolution algorithms used by Gradle and
+Maven. Let's consider Gradle first because it is the simpler one to
+describe. Simply put, Gradle chooses the highest version encountered in your
+dependency graph. For the example above, Gradle picks A:10.1.
+
+<img src="assets/images/ddc-resolution-04.png" alt="Gradle example 1">
```

Comment:
alt text should be a description of the content of the image in words sufficient for a non-sighted user. These images are core parts of the article, not simple decorations, so it's worth writing this out in some detail. E.g.

D: 40.0 is the root.
It depends on B:20.0 and C:30.0.
B:20.0 and C:30.0  both depend on A:10.1.
Nothing depends on A:10.0.